javaslang/src/main/java/javaslang/collection/Seq.java

```diff
@@ -1153,4 +1149,38 @@ default boolean startsWith(Iterable<? extends T> that, int offset) {
     default Function1<Integer, T> withDefault(Function<? super Integer, ? extends T> defaultFunction) {
         return i -> (i >= 0 && i < length()) ? apply(i) : defaultFunction.apply(i);
     }
+
+    default java.util.List<T> asUnmodifiableList() { return new SeqList<>(Seq.this); }
+
+    @SuppressWarnings("unchecked")
+    class SeqList<T> implements java.util.List<T> {
+        private final Seq<T> delegate;
+        SeqList(Seq<T> delegate) { this.delegate = delegate; }
+
+        /*@formatter:off*/
+        @Override public int size()                                  { return delegate.size(); }
+        @Override public boolean isEmpty()                           { return delegate.isEmpty(); }
+        @Override public boolean contains(Object element)            { return delegate.contains((T) element); }
+        @Override public boolean containsAll(Collection<?> elements) { return delegate.containsAll((Iterable<? extends T>) elements.iterator()); }
+        @Override public java.util.Iterator<T> iterator()            { return delegate.iterator(); }
+        @Override public T get(int index)                            { return delegate.get(index); }
+        @Override public int indexOf(Object element)                 { return delegate.indexOf((T) element); }
+        @Override public int lastIndexOf(Object element)             { return delegate.lastIndexOf((T) element); }
+        @Override public java.util.List<T> subList(int from, int to) { return delegate.slice(from, to).asUnmodifiableList(); }
+        @Override public Object[] toArray()                          { return delegate.toJavaArray(); }
+        @Override public <A>  A[] toArray(A[] array)                 { return delegate.toJavaList().toArray(array); }
+        @Override public ListIterator<T> listIterator()              { return delegate.toJavaList().listIterator(); }
```

Comment:
I don't think that is possible without linear or quadratic worst case performance (i.e. imagine a reverse iteration)